python/lib/core/dmod/core/dataset.py

```diff
@@ -487,6 +488,252 @@ def unlink_to_dataset(self, dataset: Dataset) -> bool:
+    _detector_registry: Dict[DataFormat, Dict[str, Type['ItemDataDomainDetector']]] = {f: dict() for f in DataFormat}
+    """ 
+    Registered subclasses in the form of an outer dictionary mapping supported format to inner dictionaries, mapping 
+    subclass name to subclass.
+    """
+    _subclass_formats: Dict[Type['ItemDataDomainDetector'], DataFormat] = dict()
+    """ A collection to reverse the format for a particulr subclass. """
+
+    @classmethod
+    def get_data_format(cls) -> Optional[DataFormat]:
+        """
+        Get the registered data format for this subclass of :class:`ItemDataDomainDetector`, if it is registered.
+
+        Returns
+        -------
+        Optional[DataFormat]
+            The :class:`DataFormat` of this subclass, if it registered one; otherwise ``None``.
+        """
+        return cls._subclass_formats.get(cls)
+
+    @classmethod
+    def get_types_for_format(cls, data_format: DataFormat) -> Dict[str, Type['ItemDataDomainDetector']]:
+        """
+        Return subclass type registered as capable of detecting domains for data with the given data format.
+
+        Parameters
+        ----------
+        data_format: DataFormat
+            The data format of interest.
+
+        Returns
+        -------
+        Dict[str,Type['ItemDataDomainDetector']]
+            Dictionary (keyed by class name) of registered subclasses capable of detecting domains for data with the
+            given data format.
+        """
+        return cls._detector_registry.get(data_format, dict())
+
+    def __init_subclass__(cls, format_type: Optional[DataFormat] = None, name: Optional[str] = None, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._subclass_formats[cls] = format_type
+        cls._all_subclasses[cls] = cls.__name__ if name is None else name
+        if format_type is not None:
+            cls._detector_registry[format_type][cls.__name__ if name is None else name] = cls
+
+    def __init__(self, item: DataItem, item_name: Optional[str] = None, decode_format: str = 'utf-8', *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._item: DataItem = item
+        """ The data item for which to detect a domain. """
+        self._is_item_file = isinstance(self._item, Path)
+        """ Private flag of whether data item is a :class:`Path` object (that points to a file, per other checks). """
+        self._item_name = self._item.name if self._is_item_file else item_name
+        """ Name for the item; in some situations, contains important constraint metadata (e.g. catchment name). """
+        self._decode_format = decode_format
+        """ A decoder format sometimes used when reading data item in order to get metadata. """
+        if self._is_item_file and self._item.is_dir():
+            raise ValueError(f"{self.__class__.__name__} can't initialize with a directory path as its data item")
+
+
+class UniversalItemDomainDetector(ItemDataDomainDetector):
```

Comment:
Inline with some of the comments in #566, if this functionality is needed, I recommend we refactor this functionality into some kind of factory that is defined close to where it is used.

Reply:
Yeah, see my comment [here](https://github.com/NOAA-OWP/DMOD/pull/566#discussion_r1552134681) that (towards the bottom of it) discusses why I don't think a factory is feasible.   

In short, a typical factory can't pick the right detector subclass to create without a data format, and the practical solution to getting the necessary data format is by using domain detector.  So this _is_ a sort of brute-force bootstrapping factory, it just also implements the interface at the same time, since detecting the necessary domain is a by-product of how it works.